sound/soc/sof/intel/hda-ipc.c

```diff
@@ -67,6 +67,21 @@ int hda_dsp_ipc_send_msg(struct snd_sof_dev *sdev, struct snd_sof_ipc_msg *msg)
 	return 0;
 }
 
+void hda_dsp_ipc4_schedule_d0i3_work(struct sof_intel_hda_dev *hdev,
+				     struct snd_sof_ipc_msg *msg)
+{
+	struct sof_ipc4_msg *msg_data = msg->msg_data;
+
+	/* Ipc4 follows current design of ipc3 to use a delayed_work queue to set the d0i3 state */
+	if (SOF_IPC4_MSG_IS_MODULE_MSG(msg_data->primary) &&
+	    (SOF_IPC4_MSG_TYPE_GET(msg_data->primary) == SOF_IPC4_MOD_SET_DX ||
+	    SOF_IPC4_MSG_TYPE_GET(msg_data->primary) == SOF_IPC4_MOD_SET_D0IX))
```

Comment:
why do you need this check? don't we already do it in send_msg. Also dont you need to check if there are D0i3 compatible streams running at this point? Otherwise you'd always put the DSP only in D0i3 isnt it?

Reply:
in fact why can't you reuse hda_dsp_d0i3_work()?

Reply:
please the code.  move check code into one function.   I think you got that we checked the d0i3 compatible in hda_dsp_d0i3_work.  And hda_dsp_d0i3_work() is reused